File: frontend/src/types/graph.ts
export interface GraphNode {
  id: string;
  node_type: string;
  label: string;
  attributes: Record<string, unknown>;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation: string;
  attributes: Record<string, unknown>;
}

export interface GraphStatistics {
  node_count: number;
  edge_count: number;
  node_types: Record<string, number>;
  relationship_types: Record<string, number>;
}

export interface GraphImpactAnalysis {
  root_node_id: string;
  reachable_node_count: number;
  affected_nodes: GraphNode[];
  affected_relationships: string[];
  critical_paths: string[][];
}

export interface KnowledgeGraphOverview {
  statistics: GraphStatistics;
  sample_nodes: GraphNode[];
  sample_edges: GraphEdge[];
}

export interface GraphPathResult {
  path_found: boolean;
  path_nodes: GraphNode[];
  path_edges: GraphEdge[];
  explanation: string;
}

export interface GraphNeighborhoodResult {
  center: GraphNode;
  depth: number;
  nodes: GraphNode[];
  edges: GraphEdge[];
  impact_analysis: GraphImpactAnalysis;
}

export interface GraphNodeDetail {
  node: GraphNode;
  neighbors: GraphNode[];
  edges: GraphEdge[];
  impact_analysis: GraphImpactAnalysis;
}

File: backend/app/schemas/graph.py
from typing import Any

from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    id: str
    node_type: str
    label: str
    attributes: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    source: str
    target: str
    relation: str
    attributes: dict[str, Any] = Field(default_factory=dict)


class GraphStatistics(BaseModel):
    node_count: int
    edge_count: int
    node_types: dict[str, int] = Field(default_factory=dict)
    relationship_types: dict[str, int] = Field(default_factory=dict)


class GraphImpactAnalysis(BaseModel):
    root_node_id: str
    reachable_node_count: int
    affected_nodes: list[GraphNode] = Field(default_factory=list)
    affected_relationships: list[str] = Field(default_factory=list)
    critical_paths: list[list[str]] = Field(default_factory=list)


class KnowledgeGraphOverview(BaseModel):
    statistics: GraphStatistics
    sample_nodes: list[GraphNode] = Field(default_factory=list)
    sample_edges: list[GraphEdge] = Field(default_factory=list)


class GraphNodeDetail(BaseModel):
    node: GraphNode
    neighbors: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    impact_analysis: GraphImpactAnalysis


class GraphPathResult(BaseModel):
    path_found: bool
    path_nodes: list[GraphNode] = Field(default_factory=list)
    path_edges: list[GraphEdge] = Field(default_factory=list)
    explanation: str


class GraphNeighborhoodResult(BaseModel):
    center: GraphNode
    depth: int
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    impact_analysis: GraphImpactAnalysis
